File: library/src/main/java/me/shkschneider/skeleton/data/CsvReader.java
package me.shkschneider.skeleton.data;

import android.support.annotation.Nullable;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.Reader;
import java.util.ArrayList;
import java.util.List;

import me.shkschneider.skeleton.helper.LogHelper;

// <http://opencsv.sourceforge.net>
public class CsvReader {

    private BufferedReader mBufferedReader;
    private boolean mHasNext;
    private char mSeparator;
    private char mQuote;

    public CsvReader(final Reader reader) {
        this(reader, ',', '"', 0);
    }

    public CsvReader(final Reader reader, final char separator, final char quote, final int skip) {
        mBufferedReader = new BufferedReader(reader);
        mSeparator = separator;
        mQuote = quote;
        mHasNext = true;
        if (skip > 0) {
            try {
                for (int i = 0; i < skip; i++) {
                    mBufferedReader.readLine();
                }
            }
            catch (final IOException e) {
                LogHelper.warning(e);
            }
        }
    }

    public boolean hasNext() {
        return mHasNext;
    }

    @Nullable
    public List<String> readNext() {
        try {
            final String nextLine = getNextLine();
            return parseLine(nextLine);
        }
        catch (final IOException e) {
            return null;
        }
    }

    private String getNextLine() throws IOException {
        final String nextLine = mBufferedReader.readLine();
        if (nextLine == null) {
            mHasNext = false;
            throw new IOException("EOF");
        }
        return nextLine;
    }

    @Nullable
    private List<String> parseLine(String nextLine) throws IOException {
        if (nextLine == null) {
            return null;
        }
        final List<String> tokensOnThisLine = new ArrayList<>();
        StringBuffer stringBuffer = new StringBuffer();
        boolean inQuotes = false;
        do {
            if (inQuotes) {
                // continuing a quoted section, re-append newline
                stringBuffer.append("\n");
                nextLine = getNextLine();
                if (nextLine == null) {
                    break;
                }
            }
            for (int i = 0; i < nextLine.length(); i++) {
                char c = nextLine.charAt(i);
                if (c == mQuote) {
                    // this gets complex... the quote may end a quoted block, or escape another quote.
                    // do a 1-char lookahead:
                    if (inQuotes  // we are in quotes, therefore there can be escaped quotes in here.
                            && nextLine.length() > (i + 1)  // there is indeed another character to check.
                            && nextLine.charAt(i + 1) == mQuote) { // ..and that char. is a quote also.
                        // we have two quote chars in a row == one quote char, so consume them both and
                        // put one on the token. we do *not* exit the quoted text.
                        stringBuffer.append(nextLine.charAt(i + 1));
                        i++;
                    }
                    else {
                        inQuotes = ! inQuotes;
                        // the tricky case of an embedded quote in the middle: a,bc"d"ef,g
                        if (i > 2 // not on the beginning of the line
                                && nextLine.charAt(i - 1) != mSeparator // not at the beginning of an escape sequence
                                && nextLine.length() > (i + 1)
                                && nextLine.charAt(i + 1) != mSeparator) { // not at the	end of an escape sequence
                            stringBuffer.append(c);
                        }
                    }
                }
                else if (c == mSeparator && ! inQuotes) {
                    tokensOnThisLine.add(stringBuffer.toString());
                    stringBuffer = new StringBuffer(); // start work on next token
                }
                else {
                    stringBuffer.append(c);
                }
            }
        } while (inQuotes);
        tokensOnThisLine.add(stringBuffer.toString());
        return tokensOnThisLine;

    }

    public void close() throws IOException {
        mBufferedReader.close();
    }

}

File: library/src/main/java/me/shkschneider/skeleton/data/CsvReader.kt
package me.shkschneider.skeleton.data

import java.io.BufferedReader
import java.io.IOException
import java.util.ArrayList

import me.shkschneider.skeleton.helper.LogHelper

// <http://opencsv.sourceforge.net>
class CsvReader {

    private val _bufferedReader: BufferedReader
    private val _separator: Char
    private val _quote: Char
    private var _hasNext: Boolean = false

    constructor(bufferedReader: BufferedReader, separator: Char = ',', quote: Char = '"', skip: Int = 0) {
        _bufferedReader = bufferedReader
        _separator = separator
        _quote = quote
        _hasNext = true
        if (skip > 0) {
            try {
                for (i in 0 until skip) {
                    bufferedReader.readLine()
                }
            } catch (e: IOException) {
                LogHelper.warning(e)
            }
        }
    }

    @Throws(IOException::class)
    private fun getNextLine(): String {
        val nextLine = _bufferedReader.readLine()
        if (nextLine == null) {
            _hasNext = false
            throw IOException("EOF")
        }
        return nextLine
    }

    fun hasNext(): Boolean {
        return _hasNext
    }

    fun readNext(): List<String>? {
        try {
            val nextLine = getNextLine()
            return parseLine(nextLine)
        } catch (e: IOException) {
            return null
        }
    }

    @Throws(IOException::class)
    private fun parseLine(nextLine: String): List<String>? {
        var line: String = nextLine
        val tokensOnThisLine = ArrayList<String>()
        var stringBuffer = StringBuffer()
        var inQuotes = false
        do {
            if (inQuotes) {
                // continuing a quoted section, re-append newline
                stringBuffer.append("\n")
                line = getNextLine()
            }
            var i = 0
            while (i < line.length) {
                val c = line[i]
                if (c == _quote) {
                    // this gets complex... the _quote may end a quoted block, or escape another _quote.
                    // do a 1-char lookahead:
                    if (inQuotes  // we are in quotes, therefore there can be escaped quotes in here.
                            && line.length > i + 1  // there is indeed another character to check.
                            && line[i + 1] == _quote) { // ..and that char. is a _quote also.
                        // we have two _quote chars in a row == one _quote char, so consume them both and
                        // put one on the token. we do *not* exit the quoted text.
                        stringBuffer.append(line[i + 1])
                        i++
                    } else {
                        inQuotes = !inQuotes
                        // the tricky case of an embedded _quote in the middle: a,bc"d"ef,g
                        if (i > 2 // not on the beginning of the line
                                && line[i - 1] != _separator // not at the beginning of an escape sequence
                                && line.length > i + 1
                                && line[i + 1] != _separator) { // not at the	end of an escape sequence
                            stringBuffer.append(c)
                        }
                    }
                } else if (c == _separator && !inQuotes) {
                    tokensOnThisLine.add(stringBuffer.toString())
                    stringBuffer = StringBuffer() // start work on next token
                } else {
                    stringBuffer.append(c)
                }
                i++
            }
        } while (inQuotes)
        tokensOnThisLine.add(stringBuffer.toString())
        return tokensOnThisLine
    }

    @Throws(IOException::class)
    fun close() {
        _bufferedReader.close()
    }

}
